app/src/main/java/org/wikipedia/search/SearchResultsFragment.java

```diff
@@ -347,25 +389,56 @@ private void displayResults(List<SearchResult> results) {
             for (SearchResult result : totalResults) {
                 if (newResult.getPageTitle().equals(result.getPageTitle())) {
                     contains = true;
+                    if (newResult.getPriority() == SEARCH_RESULT.getPriority()
+                            || newResult.getPriority() <= result.getPriority()) {
+                        break;
+                    }
+                    totalResults.remove(result);
+                    insertSearchResultInCorrectPosition(newResult);
                     break;
                 }
             }
             if (!contains) {
-                totalResults.add(newResult);
+                if (newResult.getPriority() == SEARCH_RESULT.getPriority() || totalResults.isEmpty()) {
+                    totalResults.add(newResult);
+                } else {
+                    insertSearchResultInCorrectPosition(newResult);
+                }
             }
         }
 
         if (totalResults.isEmpty()) {
-            searchEmptyView.setVisibility(View.VISIBLE);
-            searchResultsContainer.setVisibility(View.GONE);
+            searchEmptyView.setVisibility(VISIBLE);
+            searchResultsContainer.setVisibility(GONE);
         } else {
-            searchEmptyView.setVisibility(View.GONE);
-            searchResultsContainer.setVisibility(View.VISIBLE);
+            searchEmptyView.setVisibility(GONE);
+            searchResultsContainer.setVisibility(VISIBLE);
         }
 
         getAdapter().notifyDataSetChanged();
     }
 
+    private void insertSearchResultInCorrectPosition(@NonNull SearchResult newResult) {
```

Comment:
Not sure about the purpose of reshuffling the search results... If we make sure to put them in the correct order in the initial `prefixSearch` response, do we need to worry about reordering them again?

Reply:
@dbrant this only looks at the first 3 spots, and only for non search items. Inserting the tab, history and list item at their correct place will not work always because of cases like having to delete the less priority duplicate, a new item, say history,  getting added after a visit to the article page and back to search, or having only one of those 3 types of items, etc.